java/arcs/core/util/Scheduler.kt

```diff
@@ -11,141 +11,135 @@
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
 import kotlinx.coroutines.suspendCancellableCoroutine
 import kotlinx.coroutines.withTimeout
 
+@Suppress("UNUSED_PARAMETER")
+@Deprecated(
+    "Time is no longer of the essence",
+    ReplaceWith("Scheduler(context, agendaProcessingTimeoutMs)")
+)
+fun Scheduler(
+    time: Time,
+    context: CoroutineContext,
+    agendaProcessingTimeoutMs: Long = DEFAULT_AGENDA_PROCESSING_TIMEOUT_MS
+): Scheduler = Scheduler(context, agendaProcessingTimeoutMs)
+
 /**
  * The [Scheduler] is responsible for scheduling the execution of a batch of [Task]s (known as an
- * [Agenda]), at a defined maximum [scheduleRateHz], where each agenda is limited to running within
- * the specified [agendaProcessingTimeoutMs] window.
+ * [Agenda]).  Groups of [Task]s sent to [schedule] are guaranteed to be run in the order in which
+ * they were received, while the group's order itself is defined by a natural ordering of
+ * [Processor]s followed by grouped [Listener]s.
  */
-@Suppress("EXPERIMENTAL_API_USAGE")
+@Suppress("EXPERIMENTAL_API_USAGE", "MemberVisibilityCanBePrivate")
 class Scheduler(
-    private val time: Time,
     context: CoroutineContext,
-    private val agendaProcessingTimeoutMs: Long = DEFAULT_AGENDA_PROCESSING_TIMEOUT_MS,
-    private val scheduleRateHz: Int = DEFAULT_SCHEDULE_RATE_HZ
+    private val agendaProcessingTimeoutMs: Long = DEFAULT_AGENDA_PROCESSING_TIMEOUT_MS
 ) {
     private val log = TaggedLog { "Scheduler(${hashCode()})" }
     /* internal */
     val launches = atomic(0)
-    /* internal */
-    val loops = atomic(0)
     val scope = CoroutineScope(context)
 
-    private val isActive: Boolean
-        get() = !(processingJob?.isCompleted ?: true)
-    private var processingJob: Job? = null
-    private val isPaused = atomic(false)
-    private val agenda = atomic(Agenda())
+    private val agendaChannel = Channel<Agenda>(Channel.UNLIMITED)
+    private val pausedChannel = ConflatedBroadcastChannel(false)
+    private val idlenessChannel = ConflatedBroadcastChannel(true)
+
+    /**
+     * Flow of booleans indicating when the Scheduler has entered (`true`) or exited (`false`) and
+     * idle state.
+     */
+    val idlenessFlow: Flow<Boolean> = idlenessChannel.asFlow()
+
+    init {
+        // Consume the agenda channel:
+        // 1. Wait until the latest pause-value is false,
+        // 2. Notify the idleness channel that we're busy,
+        // 3. Do the work
+        // 4. Notify the idleness channel that we're not busy.
+        agendaChannel.consumeAsFlow()
+            .onEach { agenda ->
+                pausedChannel.asFlow().filter { !it }.first()
+                launches.incrementAndGet()
+
+                idlenessChannel.send(false)
+                try {
+                    withTimeout(agendaProcessingTimeoutMs) { executeAgenda(agenda) }
+                } catch (e: Throwable) {
+                    throw e
```

Comment:
log?

Reply:
re-throwing will also cause a log to happen

Reply:
Probably worth a comment then; have a catch-and-just-rethrow is likely to be confusing for readers (I missed this earlier).

Reply:
I could just drop it, and only use the `finally`...  would be cleaner that way.